distribution/lib/Standard/Base/0.0.0-dev/src/Data/Json/Extensions.enso

```diff
@@ -1,136 +1,132 @@
-   object key and the value as the object value.
-
-   > Example
-     Convert a vector to JSON.
-         [1, 2, 3, 4].to_json
-Any.to_json : Json
-Any.to_json self =
+## Converts the given value to a JSON serialized value.
+Any.to_json : Text
+Any.to_json self = Json.stringify self
+
+## Converts the given value to a JSON serialized value.
+Error.to_json : Text
+Error.to_json self = self.to_js_object.to_text
+
+## Converts the given value to a JSON serializable object.
+   For Nothing, booleans, numbers and strings, this is the value itself.
+   For arrays or vectors, the elements are converted recursively.
+   For atoms and maps, the object is converted to a JS_Object.
+Text.to_js_object : JS_Object | Text
+Text.to_js_object self = case self of
+    Text -> JS_Object.from_pairs [["type", "Text"]]
+    _ -> self
+
+## Converts the given value to a JSON serializable object.
+   For Nothing, booleans, numbers and strings, this is the value itself.
+   For arrays or vectors, the elements are converted recursively.
+   For atoms and maps, the object is converted to a JS_Object.
+Number.to_js_object : JS_Object | Number
+Number.to_js_object self = case self of
+    Number -> JS_Object.from_pairs [["type", "Number"]]
+    Integer -> JS_Object.from_pairs [["type", "Integer"]]
+    Decimal -> JS_Object.from_pairs [["type", "Decimal"]]
+    _ -> self
+
+## Converts the given value to a JSON serializable object.
+   For Nothing, booleans, numbers and strings, this is the value itself.
+   For arrays or vectors, the elements are converted recursively.
+   For atoms and maps, the object is converted to a JS_Object.
+Boolean.to_js_object : JS_Object | Boolean
+Boolean.to_js_object self = case self of
+    Boolean -> JS_Object.from_pairs [["type", "Boolean"]]
+    _ -> self
+
+## Converts the given value to a JSON serializable object.
+   For Nothing, booleans, numbers and strings, this is the value itself.
+   For arrays or vectors, the elements are converted recursively.
+   For atoms and maps, the object is converted to a JS_Object.
+Nothing.to_js_object : Nothing
+Nothing.to_js_object self = self
+
+## Converts the given value to a JSON serializable object.
+   For Nothing, booleans, numbers and strings, this is the value itself.
+   For arrays or vectors, the elements are converted recursively.
+   For atoms and maps, the object is converted to a JS_Object.
+Array.to_js_object : Vector
+Array.to_js_object self =
+    proxy = Array_Proxy.new self.length i-> self.at i . to_js_object
+    Vector.from_polyglot_array proxy
+
+## Converts the given value to a JSON serializable object.
+   For Nothing, booleans, numbers and strings, this is the value itself.
+   For arrays or vectors, the elements are converted recursively.
+   For atoms, the object is converted to a JS_Object.
+Vector.to_js_object : Vector
+Vector.to_js_object self =
+    proxy = Array_Proxy.new self.length i-> self.at i . to_js_object
+    Vector.from_polyglot_array proxy
```

Comment:
We are using this pattern:
```
Array_Proxy.new x.length i-> f (x.at i)
```
more and more

IMO this calls for defining a `Vector.lazy_map` encapsulating that pattern - it will make it cleaner and easier to detect usages of that pattern.

Reply:
Another topic is - whether we should use it here. The `Array_Proxy` is not _always_ a better solution.

While its construction is almost free, taking `O(1)` time, it does not come without drawbacks.
It is ideal for constructing huge collections whose elements will be read at most once, or even better - if not all elements of the collection will be touched at all.

I think we need to be careful with this lazy map pattern. It can give nice almost zero-cost conversions by deferring the actual conversion to access-time. But for it to be safe we need to ensure that the `f` is `O(1)` and rather cheap - as now we pay it on _every_ access. If by any chance `f` is more costly - we now do not pay this cost once on construction but on each call -  which can be terrible for performance.

I started writing a long paragraph on why this pattern is bad for this use-case. Then I realised that it is actually OK, because as we are doing _all_ conversions at _all levels_ lazily, the cost of an indiviual call like this is rather small and it amortizes over the path of accessing nested fields. But we still need to be careful, as it can be easy to break stuff if we add a more costly operation there.